File: Trees/Java/Find Largest Value in Each Tree Row.java
// Problem link: https://leetcode.com/problems/find-largest-value-in-each-tree-row/description/
// Time: O(n)
// Space: O(n)

class Solution {
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if(root == null) {
            return result;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        int max, size;
        TreeNode node;
        queue.offer(root);

        while(!queue.isEmpty()) {
            max = Integer.MIN_VALUE;
            size = queue.size();

            for(int index = 0; index < size; index++) {
                node = queue.poll();
                max = Math.max(node.val, max);

                if(node.left != null) {
                    queue.offer(node.left);
                }

                if(node.right != null) {
                    queue.offer(node.right);
                }
            }

            result.add(max);
        }

        return result;
    }
}

File: Trees/Python/Find Largest Value in Each Tree Row.py
# Problem link: https://leetcode.com/problems/find-largest-value-in-each-tree-row/description/
# Time: O(n)
# Space: O(n)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        result = []

        if not root:
            return result

        queue = deque([root])
        
        while queue:
            maximum = float('-inf')
            for _ in range(len(queue)):
                node = queue.popleft()
                maximum = max(maximum, node.val)

                if node.left:
                    queue.append(node.left)
                
                if node.right:
                    queue.append(node.right)
            result.append(maximum)

        return result
